pkg/cloud/rgraph/algo/parallel_queue.go

```diff
@@ -0,0 +1,241 @@
+*/
+
+package algo
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+type config struct {
+	workerCount int
+	tracer      Tracer
+}
+
+type QueueOption func(*config)
+
+func WorkerCount(n int) QueueOption  { return func(c *config) { c.workerCount = n } }
+func UseTracer(t Tracer) QueueOption { return func(c *config) { c.tracer = t } }
+
+type Tracer func(RunInfo)
+
+// RunInfo records the details of a task.
+type RunInfo struct {
+	// ID for the execution.
+	ID string
+	// Queued is the timestamp of when the task was queued.
+	Queued time.Time
+	// Start is the timestamp of when the task was started as a goroutine.
+	Start time.Time
+	// End is when the task was finished.
+	End time.Time
+	// Err is the result of the task.
+	Err error
+}
+
+// NewParallelQueue returns a new queue instance.
+func NewParallelQueue[T fmt.Stringer](opts ...QueueOption) *ParallelQueue[T] {
+	cfg := config{
+		workerCount: 2,
+		tracer:      func(RunInfo) {},
+	}
+	for _, o := range opts {
+		o(&cfg)
+	}
+	return &ParallelQueue[T]{
+		c:    cfg,
+		in:   make(chan struct{}, 100),
+		done: make(chan RunInfo, 100),
+	}
+}
+
+// ParallelQueue executes parallel tasks on separate goroutines.
+type ParallelQueue[T fmt.Stringer] struct {
+	c config
+
+	lock    sync.Mutex
+	state   queueState
+	pending []queueElement[T]
+	in      chan struct{}
+	done    chan RunInfo
+	active  int
+}
+
+type queueElement[T fmt.Stringer] struct {
+	ri   RunInfo
+	item T
+}
+
+type queueState int
+
+const (
+	stateNotStarted queueState = iota
+	stateRunning
+	stateDone
+)
+
+// Add an item to the queue. This method is threadsafe within op() and
+// can be called during Run(). It is NOT safe to call Add() from
+// a different, unassociated thread.
+func (q *ParallelQueue[T]) Add(item T) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	qe := queueElement[T]{
+		ri: RunInfo{
+			ID:     item.String(),
+			Queued: time.Now(),
+		},
+		item: item,
+	}
+	q.pending = append(q.pending, qe)
+	if q.state == stateRunning {
+		// Add() will always result in a call to launch() as
+		// we enqueue one element for each call to Add().
+		// This means the Run() loop will wake up AT LEAST
+		// once per `item`.
+		//
+		// `item` will in q.pending during launch() because
+		// the <-q.in will happen AFTER append(q.pending).
+		q.in <- struct{}{}
+	}
+}
+
+// Run the queue using op() to process each task. Different op()s must
+// not have interdependencies (e.g. op1 is blocked on op2 completing),
+// as this can result in a deadlock.
+//
+// Queue execution will stop if op() returns an error or the context
+// is canceled. When an error occurs, there may be goroutines that are
+// continuing to execute. Use q.WaitForOrphans() to wait for them
+// remaining goroutines.
+func (q *ParallelQueue[T]) Run(ctx context.Context, op func(context.Context, T) error) error {
+	q.lock.Lock()
+	if q.state != stateNotStarted {
+		return fmt.Errorf("Run() can only be called once (state=%d)", q.state)
```

Comment:
release lock?

Reply:
fixing -- the change merged, so this will be a follow up PR.